### PR TITLE
drop magic_kwargs, dock_kwargs from provide_function_widget

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
Following discussion in #2150 this would drop magic_kwargs from `viewer.window.add_function_widget` and magic_kwargs and dock_kwargs from the `provide_function_widget` hookspec. We's use from default values for magic_kwargs that make sense in the context of napari and functions to try and meet most use cases. I've moved the examples over to using `Annotated` to type function args when additional information was needed. This PR does not do any of the read-only stuff discussed in #2150. Also note that everything related to `viewer.window.add_dock_widget` and the `provide_dock_widget` hookspec is unchanged by this PR.

I'm very curious @tlambert03 if this hookspec would still be usable for the "skimage-filters" plugin you mentioned here https://napari.zulipchat.com/#narrow/stream/212875-general/topic/scikit-image.20filters.20widgets/near/223724546 (or if it could be made so by suitable changes of defaults). If not then that's an indication we're moving in the wrong direction. If so then I think we might be moving in the right direction!

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/plugins/hook_specifications.py
Comment (tlambert03) [inline napari/plugins/hook_specifications.py]: maybe we should take out the `_widget` part of the name now?

Comment (sofroniewn) [reply napari/plugins/hook_specifications.py]: Makes sense. Should we just go with `provide_function` or `provide_analysis_function`? I'm open to either.

**Comment by liu-ziyang** [comment]
I am concerned because I am not sure what feedback we would get from plugin developers. On one hand, we want to finalize the API format so they can have a solid start without getting confused about different version, on the other hand, I felt we should ask for their opinion on whether it makes sense to expose the magic_kwargs to allow customization, there has been some GUI ask from alfa cohort devs

**Comment by tlambert03** [comment]
> I felt we should ask for their opinion on whether it makes sense to expose the magic_kwargs to allow customization, there has been some GUI ask from alfa cohort devs

just want to re-emphasize the point in https://github.com/napari/napari/issues/2150 ... devs can _still_ do this using the `dock_widget` hookspec, and magicgui's `@magic_factory`.  So what is being removed here is really duplicate functionality... but by leaving those kwargs in there, it makes future development of this very valuable hookspec difficult.

**Comment by sofroniewn** [comment]
> So what is being removed here is really duplicate functionality... but by leaving those kwargs in there, it makes future development of this very valuable hookspec difficult.

Agreed, I see this now more as removing duplicate functionality - see even https://github.com/tlambert03/proto-skimage-widgets/pull/1 for what it's like to use this hookspec on the `skimage-filters` with now `magic_kwargs` or `dock_kwargs`.

We still have our more general dock hookspec so any plugin developer can always use that if they want

File: napari/plugins/hook_specifications.py
Comment (liu-ziyang) [reply napari/plugins/hook_specifications.py]: `provide_funtion` sounds better to me

**Comment by liu-ziyang** [comment]
> > I felt we should ask for their opinion on whether it makes sense to expose the magic_kwargs to allow customization, there has been some GUI ask from alfa cohort devs
> 
> just want to re-emphasize the point in #2150 ... devs can _still_ do this using the `dock_widget` hookspec, and magicgui's `@magic_factory`. So what is being removed here is really duplicate functionality... but by leaving those kwargs in there, it makes future development of this very valuable hookspec difficult.

If we can get broad adoption of the annotation-based registration already that would be awesome, if we are determined this is the right way, we should do it quickly and add demo usage accordingly. Right now I am confused because I don't see the magic gui annotation anywhere to register gui component for the existing examples

**Comment by sofroniewn** [comment]
> Right now I am confused because I don't see the magic gui annotation anywhere to register gui component for the existing examples

The main change in this PR is that this hookspec now forces users to put parameters in the function signature instead of the `magic_kwargs` and they loose access to the "global" `magic_kwargs` like `layout`, napari now uses it's own defaults, which I think is really nice.

```python
def gaussian_blur(
    layer: 'napari.layers.Image',
    sigma: Annotated[float, {"widget_type": "FloatSlider", "max": 6}] = 1.0,
    mode: Annotated[str, {"choices": ["reflect", "constant", "nearest", "mirror", "wrap"]}]="nearest",
) -> 'napari.types.ImageData':
    """Apply a gaussian blur to ``layer``."""
    if layer:
        return skimage.filters.gaussian(layer.data, sigma=sigma, mode=mode)
```

One thing to think about bigger picture is what it's like for one hookspec to use the functions provided by this hookspec. For example @haesleinhuepf I'd love to see the operations in https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_operations/_operations.py be provided by this hookspec with no qt/ magicgui and then the assistent be able to leverage those (and possibly other functions provided by this hookspec https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_gui/_Assistant.py) Does that make any sense?

@tlambert03 @jni is there any more you'd want to see from this before final review aside from the name change?

We could add `napari_experimental_provide_function` and deprecate `napari_experimental_provide_function_widget`? That is maybe the safest option - if people are using the deprecated one and then don't like the new one they can maybe tell us before it is forever taken away!! I might rewrite the PR to look more like a deprecation than a breaking change. Thoughts?

File: examples/magic_parameter_sweep.py
Comment (tlambert03) [inline examples/magic_parameter_sweep.py]: probably should specify that the keys of these dicts are magicgui widget kwargs, so they know what's available

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: i often feel like `auto_call` and `call_button` are weird together, so we should probably pick one for now.  This is likely the hardest one for us to "guess" about.  Though it's no fun for short running functions, using `auto_call` can stink for long running functions.  So i think we might want to use `call_button` for now, and think about how we could pick between them (or sneak it into the annotation somehow)

File: napari/plugins/__init__.py
Comment (tlambert03) [inline napari/plugins/__init__.py:123]: something I recently came across is that checking `callable(func)` is not the same as `isinstance(func, types.FunctionType)` ... and arbitrary class instances that have a `__call__` method (but not a `__name__`) will pass the first check throw an `AttributeError` here.  So we should either use `isinstance(func, types.FunctionType)` above, or use `name = getattr(func, '__name__', 'Unnamed Callable').replace('_', ' ')` here

**Review by tlambert03 — APPROVED** [review]
> @tlambert03 @jni is there any more you'd want to see from this before final review aside from the name change?

nope this looks good to me!  just the name change

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: We can default to `call_button` for now, it is safer

File: examples/magic_parameter_sweep.py
Comment (sofroniewn) [reply examples/magic_parameter_sweep.py]: good point

Comment (sofroniewn) [reply examples/magic_parameter_sweep.py]: is there a magicgui link i can put in?

File: napari/plugins/__init__.py
Comment (sofroniewn) [reply napari/plugins/__init__.py:123]: I'd probably vote  `isinstance(func, types.FunctionType)` and use `types.FunctionType` everywhere as that seems more in line with what we want. Thoughts?

File: examples/magic_parameter_sweep.py
Comment (tlambert03) [reply examples/magic_parameter_sweep.py]: sure, just say "keywords must be valid for the corresponding magicgui widget type", and link to https://napari.org/magicgui/api/widgets.html

File: napari/plugins/__init__.py
Comment (tlambert03) [reply napari/plugins/__init__.py:123]: it's safer, we may be excluding valid clever uses for now, but we can open it up further if the need arises.

**Comment by sofroniewn** [comment]
Ok, I've made requested changes. I've decided against deprecation here, as it is only one release for something marked as `experimental`. I have also updated our plugin documentation to close #2126 - @tlambert03 can you check that please.

Maybe @jni can give a final thumbs up and then we can release this and cut the 0.4.4 release as discussed in the zulip

**Review by tlambert03 — APPROVED** [review]

**Comment by jni** [comment]
@sofroniewn @tlambert03 

re deprecation, I think it would still be nice to have, even if it's a hard one (ie raise an error message, but rather than a generic one, raise one that points users in the right direction). If anyone has been using the function hookspec till now, it would at least be nice for napari to raise an error with a message about what has been deprecated and how to migrate. It might save some brave soul a bit of searching around.

tbh the *motivation* for this PR is still unclear to me. We want a consistent interface for all functional plugins? That's fine, I like that well enough. But it's not very clearly stated above. Someone might have to do a lot of digging around GH discussions to figure it out.

The final concern I have is that I have no idea how to override auto_call=True. If I have a function that takes 25min to run I probably want to set auto_call=False! Is that at all possible with Annotated?

Other than the concerns above I do like the changes — returning a list of functions is much nicer and clearer!

**Comment by jni** [comment]
> tbh the motivation for this PR is still unclear to me.

One partial way to address this is for whoever merges this to make sure that the commit message clearly explains the motivation.

**Comment by tlambert03** [comment]
> tbh the _motivation_ for this PR is still unclear to me. We want a consistent interface for all functional plugins?

for me it's less about the ui of plugins, and more about what I tried to lay out in https://github.com/napari/napari/issues/2150#issue-792318998.  The point there was that we currently have overlapping functionality in the two hookspecs, and if we have a vision of someday "automagically" turning well-annotated functions into graphical workflows (including pipelines), and if _this_ is going to be the experimental hookspec for that, then we should not get anyone used to having magicgui-like control of those elements (outside of parameter annotations) until it's clearer to us how that will be used.  (for instance, if we _do_ have a node-based editor then these metakwargs like `call_button` or `auto_run` make a lot less sense).  Meanwhile, anyone who would like to take advantage of magicgui's ability to turn their functions into widgets (complete with auto_call, etc...) can do so using `@magic_factory` and `provide_dock_widget`. 


> The final concern I have is that I have no idea how to override auto_call=True. If I have a function that takes 25min to run I probably want to set auto_call=False! Is that at all possible with Annotated?

I think @sofroniewn  changed this? https://github.com/napari/napari/pull/2158#discussion_r564844498


> Other than the concerns above I do like the changes — returning a list of functions is much nicer and clearer!

**Comment by jni** [comment]
> > The final concern I have is that I have no idea how to override auto_call=True. If I have a function that takes 25min to run I probably want to set auto_call=False! Is that at all possible with Annotated?
> 
> I think @sofroniewn changed this?

Oh right! Well, then the question is the converse? How do we give @haesleinhuepf the possibility to have auto_call turned on for his blazing fast OpenCL code?

One idea, @tlambert03: how about we add a tick box "auto-call" or "run automatically" to the magicgui-generated output by default? Then at least the user clicks this one thing and from then on, the output is automatically updated?

But yeah, ok, based on that suggestion, whether or not we take it, I see how we can solve this issue with our autogeneration, rather than by giving users control over these options.

Merge at will, @sofroniewn! 😊

**Comment by tlambert03** [comment]
> But yeah, ok, based on that suggestion, whether or not we take it, I see how we can solve this issue with our autogeneration, rather than by giving users control over these options.

yep, this exactly.  If they need control, use `provide_dock_widget` with `@magic_factory` ... it's all there for them

**Comment by jni** [comment]
> yep, this exactly. If they need control, use provide_dock_widget with @magic_factory ... it's all there for them

Well, the point though is that functions should go in functions. e.g. we don't want pyclesperanto to be in its own widget playland, we want to be able to compose it with all the other functions we will provide.

**Comment by tlambert03** [comment]
> > yep, this exactly. If they need control, use provide_dock_widget with @magic_factory ... it's all there for them
> 
> Well, the point though is that functions should go in functions. e.g. we don't want pyclesperanto to be in its own widget playland, we want to be able to compose it with all the other functions we will provide.

we're on the same page here @jni.  That's what this hookspec is for.  functions should be functions.  not their own widget playland.  But we literally don't have a single place in the codebase where we "apply" functions as functions :joy:!  They're _only_ widgets at the moment.  I'm just saying lets think for a moment about how we might either offer up that functional api without accepting magic_kwargs... or, if this all comes down to the decision between `auto_call` and `call_button`, I'm sure we can figure something out without exposing all of the magicgui API in `provide_function` (particularly when they have that, _for now_, at their disposal in `provide_dock_widget`).\

**Comment by sofroniewn** [comment]
Ok, thanks for the additional discussion @tlambert03 @jni - I have explicitly made `auto_call=False`. I've decide to leave `magic_kwargs` as an input to `viewer.window.add_function_widget` which is explicitly about a function widget. I think it's ok to drop and rename the initial hookspec so soon after release, it would be a big deal to deprecate nicely, and it is in `experimental` for a reason - a different deprecation policy applies. I will merge this after tests pass

**Pull request merged**
